### PR TITLE
Sane detection of the host/port in entrypoint prod

### LINKED ISSUE #17828: Docker image entrypoint does not parse redis URLs correctly with no password
**Apache Airflow version**:

2.1.3

**Deployment**:

ECS, docker-compose

**What happened**:

A redis connection URL of the form: 

```
redis://<host>:6379/1
```

is parsed incorrectly by this pattern: https://github.com/apache/airflow/blob/2.1.3/scripts/in_container/prod/entrypoint_prod.sh#L96 - the host is determined to be `localhost` as there is no `@` character.

Adding an `@` character, e.g. `redis://@redis-analytics-airflow.signal:6379/1` results in correct parsing, but cannot connect to redis as auth fails.

Logs:

```
BACKEND=redis
DB_HOST=localhost
DB_PORT=6379
```

and eventually

```
ERROR! Maximum number of retries (20) reached.
Last check result:
$ run_nc 'localhost' '6379'
(UNKNOWN) [127.0.0.1] 6379 (?) : Connection refused
sent 0, rcvd 0
```

**How to reproduce it**:

Run airflow with an official docker image and `  AIRFLOW__CELERY__BROKER_URL: redis://<ANY_REDIS_HOSTNAME>:6379/1` and observe that it parses the host as `localhost`.

**Anything else we need to know**:

This was working without issue in Airflow 2.1.2 - weirdly I cannot see changes to the entrypoint, but we have not changed the URL `AIRFLOW__CELERY__BROKER_URL: redis://redis-analytics-airflow.signal:6379/1`.

### PR DESCRIPTION
**Pull request by potiuk**
The previous regexp parsing was well, not perfect closely following the
ancient Chinese proverb "If you have problem, introduce regexp - you
will have two problems".

This PR replaces regexp matching with python urlsplit method.

Close: https://github.com/apache/airflow/issues/17828

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: scripts/in_container/prod/entrypoint_prod.sh [hunk 1]
```diff
@@ -83,24 +84,12 @@ function wait_for_connection {
...
-    [[ ${connection_url} =~ ([^:]*)://([^:@]*):?([^@]*)@?([^/:]*):?([0-9]*)/([^\?]*)\??(.*) ]] && \
-        detected_backend=${BASH_REMATCH[1]} &&
-        # Not used USER match
-        # Not used PASSWORD match
-        detected_host=${BASH_REMATCH[4]} &&
-        detected_port=${BASH_REMATCH[5]} &&
-        # Not used SCHEMA match
-        # Not used PARAMS match
+    local detected_backend
+    detected_backend=$(python -c "from urllib.parse import urlsplit; print(urlsplit('${connection_url}').scheme)")
```
Comment (mik-laj) [inline scripts/in_container/prod/entrypoint_prod.sh]:
```suggestion
    detected_backend=$(python -c "from urllib.parse import urlsplit; import sys; print(urlsplit(sys.argv[1]).scheme)" "${connection_url}")
```
I am concerned that some inputs may lead to Command Injection and therefore incorrect script behavior.

Comment (mik-laj) [reply scripts/in_container/prod/entrypoint_prod.sh]: For readability, we can extraactt it to a separate function.
```bash
function get_url_part() {
    url_part="$1";
    connection_url="$2";
    python -c "from urllib.parse import urlsplit; import sys; print(getattr(urlsplit(sys.argv[2]), sys.argv[1]))" "${url_part}" "${connection_url}"
}
get_url_part scheme redis://
```

Comment (potiuk) [reply scripts/in_container/prod/entrypoint_prod.sh]: I prefer to leave it explicit. I think it is far less readable with url_part parameter to be honest.

Comment (potiuk) [reply scripts/in_container/prod/entrypoint_prod.sh]: adding separate function is what I tried initially, and I REALLY found it unnecessary

Comment (potiuk) [reply scripts/in_container/prod/entrypoint_prod.sh]: > I am concerned that some inputs may lead to Command Injection and therefore incorrect script behavior.

Yep. Agree on that one.

**Review by mik-laj — APPROVED** [review]

**Pull request merged**